File: python/medium/Binary_Tree_Level_Order_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        ans = [[]]
        def next_node(node, layer):
            if len(ans) < layer:
                ans.append([])
            ans[layer-1].append(node.val)
            if node.left is not None:
                next_node(node.left, layer+1)
            if node.right is not None:
                next_node(node.right, layer+1)
        next_node(root, 1)
        return ans

File: cpp/medium/Binary_Tree_Level_Order_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (root != nullptr) next_node(root, ans, 0);
        return ans;
    }
    void next_node(TreeNode* node, vector<vector<int>>& ans, int level) {
        if (ans.size() <= level) {
            vector<int> comb{node->val};
            ans.push_back(comb);
        }
        else ans[level].push_back(node->val);
        if(node->left != nullptr) next_node(node->left, ans, level+1);
        if(node->right != nullptr) next_node(node->right, ans, level+1);
    }
};
